src/common/mmap.c

```diff
@@ -492,47 +501,100 @@ util_range_unregister(const void *addr, size_t len)
 }
 
 /*
- * util_range_is_pmem -- return true if entire range
- * is persistent memory
+ * util_range_is_pmem -- return true if entire range is persistent memory
  */
 int
-util_range_is_pmem(const void *addrp, size_t len)
+util_range_is_pmem(const void *addr, size_t len)
 {
-	LOG(10, "addr %p len %zu", addrp, len);
+	LOG(10, "addr %p len %zu", addr, len);
 
-	uintptr_t addr = (uintptr_t)addrp;
-	int retval = 1;
+	if (len > UINTPTR_MAX - (uintptr_t)addr) {
+		len = UINTPTR_MAX - (uintptr_t)addr;
+		LOG(4, "limit len to %zu to not get beyond address space", len);
+	}
+
+	uintptr_t begin = (uintptr_t)addr;
+	uintptr_t end = (uintptr_t)addr + len;
+
+	LOG(4, "begin 0x%016" PRIxPTR " end 0x%016" PRIxPTR, begin, end);
+
+	int retval = -1;
 
 	util_rwlock_rdlock(&Mmap_list_lock);
 
-	do {
-		struct map_tracker *mt = util_range_find(addr, len);
-		if (mt == NULL) {
-			LOG(4, "address not found 0x%016" PRIxPTR, addr);
+	struct map_tracker *mt;
+	SORTEDQ_FOREACH(mt, &Mmap_list, entry) {
+		if (mt->end_addr <= begin) {
+			LOG(4, "skipping all mapped ranges before given range");
+			continue;
+		}
+		if (mt->base_addr >= end) {
+			LOG(4, "ignoring all mapped ranges beyond given range");
+			break;
+		}
+
+		if (begin < mt->base_addr) {
+			LOG(4, "untracked address");
 			retval = 0;
 			break;
 		}
 
-		LOG(10, "range found - begin 0x%016" PRIxPTR
-				" end 0x%016" PRIxPTR,
+#ifdef _WIN32
+		/*
+		 * This is only needed for Windows, where all the mappings
+		 * are tracked. On Linux, only DAX mappings are on the list.
```

Comment:
Maybe we should change this logic so that on all platforms every mapping in on the list?